presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateProjectOnValues.java

```diff
@@ -0,0 +1,150 @@
+import static com.facebook.presto.sql.planner.plan.Patterns.values;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This optimizer looks for ProjectNode followed by a ValuesNode and get the ProjectNode Evaluated.
+ * When this rule is used on iterative optimizer, the rule could apply iteratively.
+ * <p/>
+ * Plan before optimizer:
+ * <pre>
+ * ProjectNode (outputVariables)
+ *   - ValuesNode
+ * </pre>
+ * <p/>
+ * Plan after optimizer:
+ * <pre>
+ * ValuesNode (outputVariables)
+ * </pre>
+ */
+public class EvaluateProjectOnValues
+        implements Rule<ProjectNode>
+{
+    private static final Pattern<ProjectNode> PATTERN = project().with(source().matching(values()));
+
+    private final Metadata metadata;
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public EvaluateProjectOnValues(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "Metadata is null");
+        this.functionAndTypeManager = requireNonNull(metadata.getFunctionAndTypeManager(), "functionManager is null");
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isEvaluateProjectOnValues(session);
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        Optional<PlanNode> optionalSource = context.getLookup().resolveGroup(projectNode.getSource()).findFirst();
+        if (!optionalSource.isPresent()) {
+            return Result.empty();
+        }
+        ValuesNode source = (ValuesNode) optionalSource.get();
+        List<List<RowExpression>> rows = source.getRows();
+        List<VariableReferenceExpression> valuesOutputVariables = source.getOutputVariables();
+        List<VariableReferenceExpression> projectOutputVariables = projectNode.getOutputVariables();
+        List<RowExpression> projectRowExpressions = projectNode.getAssignments().getExpressions().stream().collect(toImmutableList());
+        // exclude non-deterministic function
+        DeterminismEvaluator determinismEvaluator = new RowExpressionDeterminismEvaluator(functionAndTypeManager);
+        if (!projectRowExpressions.stream().allMatch(determinismEvaluator::isDeterministic)) {
+            return Result.empty();
+        }
+        if (!rows.stream().allMatch(row -> row.stream()
+                .allMatch(determinismEvaluator::isDeterministic))) {
+            return Result.empty();
+        }
+        ExpressionOptimizer expressionOptimizer = new RowExpressionOptimizer(metadata);
+
+        //rewrite ProjectNode assignment expressions
+        List<List<RowExpression>> rowExpressionsList = rows.stream().map(row -> {
```

Comment:
I see we're operating over lists of lists for the following three blocks.  Can the outer list be converted to a for loop, and within this loop we operate over a single list?  I.e. can we unnest this?  It would simplify this code.

Reply:
fixed